fix(UI): Script displayed in preview when using specific css

Othniel01 opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #55861 

<!-- Feel free to add any additional description of changes below this line -->

This pull request introduces enhancements to the createHeader function, which now injects CSS to ensure that specific head elements (including title, meta, link, and script tags) are hidden in the preview frame. This change addresses a previously identified issue where these elements were being rendered in the preview, potentially causing confusion for users.

Created a jest test file and ran a test to ensure that the css to hide head elements and scripts is being injected, here is a screenshot below

![jest](https://github.com/user-attachments/assets/1c78f902-a175-497f-938a-c2d2f2853148)

This is the main change to the frame.ts file

![yum](https://github.com/user-attachments/assets/9709b1d7-8d23-43f7-96f2-d6fe36d2740e)

Now when adding css to the javascript and algorithm data structure beta, the scripts texts from head and body arent rendered anymore in preview

![haq](https://github.com/user-attachments/assets/770153ca-0441-4b4c-9e24-91b17540833d)

Othniel01 committed:
fix: script displayed in preview when using specific css

Othniel01 committed:
fix: script displayed in preview when using specific css

Othniel01 commented:
I apologize for the mistake, The title has been changed now to match the pull request naming convention.

naomi-lgbt commented on iframe-preview.test.js:
First, this file doesn't seem to be in the right directory. Or any directory.

Second, I think rather than testing we inject the content, we should have an E2E test which enters the CSS that triggers the issue into the editor, and confirm that the preview window does not contain content.

Othniel01 replied:
Okay I was having a hard time finding the editor element so I can insert the script into it.
This is the code in the newly created E2e test,  e2e/iframe-script.spec.ts

import { expect, test } from '@playwright/test';
import translations from '../client/i18n/locales/english/translations.json';
import { focusEditor } from './utils/editor';

test.beforeEach(async ({ page }) => {
  await page.goto(
    '/learn/javascript-algorithms-and-data-structures-v8/build-a-cash-register-project/build-a-cash-register'
  );
});

test.describe('Challenge Preview Component', () => {
  // Test for inserting <style> tag and checking if the preview is empty
  test('should render an empty preview after inserting CSS', async ({
    page,
    isMobile
  }) => {
    await focusEditor({ page, isMobile });

    // Insert the <style> tag with display: block rule into the editor
    const styleTag = `
      <style>
      * {
        margin: 0;
        padding: 0;
        display: block;
      }
      </style>`;
    await page.keyboard.insertText(styleTag);

    if (isMobile) {
      await page
        .getByRole('tab', { name: translations.learn['editor-tabs'].preview })
        .click();
    }

    // Check that the iframe preview is empty
    const frame = page.frameLocator('#fcc-main-frame').first();

    // Make sure there are no visible elements inside the iframe's body
    await expect(frame.locator('body')).toBeEmpty();
  });
});

naomi-lgbt replied:
The E2E test covers this change. This file should be removed.

Othniel01 replied:
The e2e/iframe-script.spec.ts has been removed. If im understanding you correctly there is already another E2E test to cover this change.

hosting locally, using the problematic css does not bring up the unwanted scripts any more.

![haq](https://github.com/user-attachments/assets/836737cc-bd47-4397-9731-40f5f70f9e5c)

Othniel01 committed:
fix: made new changes to prevent script displayed in preview when using specific css

Othniel01 committed:
fix: new changes on script displayed in preview when using specific css

Othniel01 committed:
fix: removed e2e/iframe-script.spec.ts

naomi-lgbt commented:
I think you misunderstood. The E2E test file should stay. The file I commented on, which is the random `iframe-preview.test.js` in the root directory, is the one that should be removed.

Othniel01 commented:
yes i probably misunderstood. Please give me a second

Othniel01 committed:
fix: added back E2E test and deleted unwanted and displaced iframe jest test

Othniel01 commented:
> I think you misunderstood. The E2E test file should stay. The file I commented on, which is the random `iframe-preview.test.js` in the root directory, is the one that should be removed.

I have added the E2E test file back and removed the random unwanted jest file. 

![iframe ts](https://github.com/user-attachments/assets/455085c9-1018-4427-a7a3-9b919248af10)

Othniel01 commented:
This checks failing seem to be coming from another file navigation-from-last-challenge.spec.ts:61:7 on chromium and chrome 
Is that an issue ?

ojeytonwilliams reviewed (changes requested):
I'm afraid this is a problem - it's no longer possible to pass the tests for the Build a Personal Portfolio Webpage project.

What's happening is that the library we use to analyse the learner's code (style code in this case) is detecting the new style element introduced by this PR.

The bigger problem is that we have a battery of tests that should have detected this, but they didn't. I'll look into that now, since we need those to work to prevent regressions.

ojeytonwilliams commented on client/src/templates/Challenges/utils/frame.ts:
```diff
@@ -88,6 +88,11 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
 
 const createHeader = (id = mainPreviewId) => `
   <base href='' />
+  <style>
```
```suggestion
  <style class="fcc-hide-header">
```
The CSSHelper script ignores `style` elements with classes, so this should fix those bugs. However, this is a bit of a hack, so let's wait until https://github.com/freeCodeCamp/freeCodeCamp/pull/56598 is in before we test this.

Assuming that this does the job, an explanatory comment (on line 87 or so) would help since it's far from obvious what this is doing.

Othniel01 replied:
If this makes the style scripts ignored by the library, then for now this should be a temporal workaround preventing the unexpected rendering behavior in the test environments.

Othniel01 replied:
Header PR has been approved, can we test this ?

ojeytonwilliams replied:
Sure. I've set the tests going.

Othniel01 replied:
1 failing

  1) Check challenges
       Language: english
         SuperBlock: 2022/responsive-web-design
           Block: learn-html-by-building-a-cat-photo-app
             Title: Step 1
               ID: 5dc174fcf86c76b9248c6eb2
                 Check tests against solutions
                   Solution 1 must pass the tests:
     TypeError: createHeader is not a function
      at initializeTestRunner (test/test-challenges.js:671:25)
      at async getContextEvaluator (test/test-challenges.js:640:3)
      at async createTestRunner (test/test-challenges.js:591:21)
      at async Context.<anonymous> (test/test-challenges.js:539:42)



 ELIFECYCLE  Test failed. See above for more details.
 ELIFECYCLE  Command failed with exit code 1.
ERROR: "test:curriculum" exited with 1.
 ELIFECYCLE  Test failed. See above for more details.
Error: Process completed with exit code 1.

Othniel01 committed:
fix: added fcc-hide-header class as temp hack

ojeytonwilliams commented on client/src/templates/Challenges/utils/frame.ts:
```diff
@@ -86,8 +86,17 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
 
-export const createHeader = (id = mainPreviewId) => `
+// The "fcc-hide-header" class is added to ensure that the CSSHelper class ignores this style element
+// during tests, preventing CSS-related test failures. This is a temporary fix.
```
```suggestion
// during tests, preventing CSS-related test failures.
```
Nitpick: we don't have any real plans to change this.

ojeytonwilliams commented on client/src/templates/Challenges/utils/frame.ts:
```diff
@@ -86,8 +86,17 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
+// The "fcc-hide-header" class is added to ensure that the CSSHelper class ignores this style element
+// during tests, preventing CSS-related test failures. This is a temporary fix.
+
+const createHeader = (id = mainPreviewId) =>
+  `
```
```suggestion
export const createHeader = (id = mainPreviewId) => `
```
This is used by the tests, so needs to be exported.

Othniel01 committed:
Update client/src/templates/Challenges/utils/frame.ts

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Othniel01 committed:
Update client/src/templates/Challenges/utils/frame.ts

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Othniel01 committed:
fix: fetch latest changes from upstream

ojeytonwilliams commented:
Hi, @Othniel01, thanks for updating the PR. It looks like the style element is causing another test to fail. Once that's addressed, this should be good to go.

Othniel01 commented:
Thank you @ojeytonwilliams @naomi-lgbt 
Please let me know in any way I can be of more help, as I am learning a lot from this contribution.

ojeytonwilliams commented:
Thanks for the offer, @Othniel01. If you have a chance to look into the failing test, that would be a big help.

Othniel01 commented:
<h1> Reason for test failure </h1>
After taking time to review the failing test, seems the test failing is related to the check for a solution in the "learn-html-by-building-a-cat-photo-app" challenge from the "2022/responsive-web-design" super block.
Our injected code is making this test to fail, as the test is meant to check if any extensions are in use
<br></br>
<p>You appear to be using a browser extension that is modifying the page. Be sure to turn off all browser extensions.</p>
and return a value of +0 if successful but we are getting -1. 
<br></br>
The test framework might have detected our style injection modification and flagged the solution as invalid, even though the solution itself might be correct. 

Now we have added a csshelper class earlier that is meant to make the tests ignore our script, but it seems not all of them ignore the script.

ojeytonwilliams commented:
That sounds right. It should be sufficient to adjust the test so that it checks that there is one style element (the one added in this PR). If it's anything other than 1, it's likely that an extension has added more.

ojeytonwilliams commented on curriculum/i18n-curriculum:
Could you revert these changes? It's never necessary to modify the i18n-curriculum directly - that's all handled by tooling.

Othniel01 replied:
Ive reverted the changes. Is there another way to set the lengths to 1 ?

ojeytonwilliams reviewed (changes requested):
Thanks for updating those tests. Could you revert the wanted changes, then we can try CI again?

Othniel01 commented:
<h1>Reverted Pull Request Changes</h1>
@ojeytonwilliams  Now in all the versions from english to ukrainian
<br></br>
Sorry how do we then set the length to 1, i there another way to make the change?  to give room for our injected style element, any other style element asides from this would trigger this error as an extension is in use.
<br></br>
<b>assert.equal(document.querySelectorAll('style').length, 1);</b>
<br></br>
Hopefully this stops the errors.

ojeytonwilliams commented:
Sorry, I think we might be talking about different things.

The problem was that you'd changed the i18n curriculum. That was the only change that needed reverting.

Could you restore the fixes you had made to the curriculum files?

Othniel01 committed:
fix: added back changes to curriculum files

Othniel01 commented:
Okay restored back the curriculum changes and left the  i18n curriculum folder alone

Othniel01 commented:
Seems like we are resolving the failing tests a step at a time, the new failing test stems from the "2022/responsive-web-design" curriculum while working on the "learn-basic-css-by-building-a-cafe-menu" block, specifically for "Step 12."

"You should not add a new <style> tag. Add the new CSS rules to the existing <style> tag."

The error indicates that our code did not meet the test requirements, which expected us to modify an existing <style> tag rather than create a new one.

 1 failing

  1) Check challenges
       Language: english
         SuperBlock: 2022/responsive-web-design
           Block: learn-basic-css-by-building-a-cafe-menu
             Title: Step 12
               ID: 5f3477ae9675db8bb7655b30
                 Check tests against solutions
                   Solution 1 must pass the tests:
     Check next step for solution!
Test text: <p>You should not add a new <code>style</code> tag. Add the new CSS rules to the existing <code>style</code> tag.</p>
  AssertionError: Unspecified AssertionError
      at eval (eval at <anonymous> (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:123872), <anonymous>:2:1)
      at HTMLDocument.<anonymous> (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:123872)
      at u (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:26888)
      at c (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:27190)

Othniel01 commented:
This is the code responsible for checking if there are more than 1 style sheets added or being used at the "2022/responsive-web-design" curriculum while working on the "learn-basic-css-by-building-a-cafe-menu" block, specifically for "Step 12."

const hasManyStyleTags = document.querySelectorAll('style').length > 1;
assert(!hasManyStyleTags);


Since we are injecting a style sheet globally, we can increase the number to 2

const hasManyStyleTags = document.querySelectorAll('style').length > 2;
assert(!hasManyStyleTags);

Allowing for our injected style script and the given style script and not more than two. any other style added would now throw the error ignoring our injected style sheet

Othniel01 committed:
fix: increased style check length to allow 2 style scripts instead of 1

ilenia-magoni commented on curriculum/challenges/english/25-front-end-development/workshop-cafe-menu/5f3477ae9675db8bb7655b30.md:
````diff
@@ -21,7 +21,7 @@ assert(hasH1);
 
 ```js
-const hasManyStyleTags = document.querySelectorAll('style').length > 1;
+const hasManyStyleTags = document.querySelectorAll('style').length > 2;
 assert(!hasManyStyleTags);
````
```suggestion
assert.isAtMost(document.querySelectorAll('style').length, 2);
```

Othniel01 replied:
Suggested changes have been taken in to account, thank you for your cleaner, and more simplified approach.

Othniel01 committed:
fix: made use of latest suggestion the .length, 2 and assert.isAtMost

ojeytonwilliams reviewed (changes requested):
Hi, @Othniel01 I'm afraid there are more tests that were written assuming there would be one style element.

At a glance they seem to be quite brittle, so it's possible there are better ways to write them. I'll ask someone on the curriculum team to advise.

Othniel01 commented:
Hey @ojeytonwilliams, thanks for your feedback, I would be waiting for any updates concerning this issue

ilenia-magoni commented:
there are merge conflicts now

Othniel01 commented:
hope the merge conflicts have been resolved now

ilenia-magoni commented:
there are still challenges failing for the new style, @ojeytonwilliams do you have updates?

ojeytonwilliams committed:
fix(curriculum): ignore new style elem in tests

ojeytonwilliams commented:
Sorry for the delay. I pushed a couple of minor changes to fix the failing tests and it should be fine now.

ojeytonwilliams reviewed (approved):
Thanks for your patience, @Othniel01. This LGTM :+1:

ojeytonwilliams commented:
@ilenia-magoni it shouldn't be necessary. Every time you push a new commit to GitHub, it implicitly rebases the entire branch on top of the target, main in this case.

ilenia-magoni commented:
I did not mean merge conflicts, I meant to see if new challenges have issues, or things like that

ojeytonwilliams commented:
Right, but that's what just happened. When I pushed my commit, GitHub rebased this PR's commits on top of main, before triggering the actions. Any new challenges that were in main at the point I pushed will have been included in the CI tests.
